File: day5/day5.rs
use std::collections::VecDeque;
use std::fs::File;
use std::io::{self, prelude::*, BufReader};

fn main() -> io::Result<()> {
    let file = File::open("./input.txt")?;
    let reader = BufReader::new(file);
    let mut box_load = true;
    let n_stacks = 9;
    let mut crates: Vec<VecDeque<char>> = Vec::new();
    let mut instr = Vec::new();

    for _ in 0..n_stacks {
      crates.push(VecDeque::new());
    }

    for line in reader.lines() {
      if box_load {
        let row = line.unwrap().chars().collect::<Vec::<char>>();
        if *row.get(1).unwrap() as i32 == '1' as i32 {
          box_load = false;
          continue;
        }
        for i in 0..n_stacks {
          if *row.get(1+i*4).unwrap() as i32 != ' ' as i32 {
            crates[i].push_front(*row.get(1+i*4).unwrap());
          }
        }
      }  else {
        let row = line.unwrap();
        if row.len() == 0 {
          continue;
        }

        let mut inst = Vec::new();
        let sp = row.split(" ").collect::<Vec<&str>>();
        inst.push(sp[1].parse().unwrap());
        inst.push(sp[3].parse().unwrap());
        inst.push(sp[5].parse().unwrap());
        instr.push(inst);
      }
    }

    // Moving
    for inst in instr {
      // Silver star
      /*for _ in 0..inst[0] {
        let _crate = crates[inst[1] as usize - 1].pop_back().unwrap();
        crates[inst[2] as usize - 1].push_back(_crate);
      }*/
      // Gold star
      let index = crates[inst[1] as usize - 1].len() - inst[0];
      for _ in 0..inst[0] {
        let _crate = crates[inst[1] as usize - 1].remove(index).unwrap();
        crates[inst[2] as usize - 1].push_back(_crate);
      }
    }
    
    for c in crates {
      print!("{}", c[c.len() as usize - 1]);
    }
    
    println!();
    Ok(())
}

File: day5/day5.kt
import java.io.File
import kotlin.streams.asSequence

fun main() {
    var lines = File("input.txt").readLines()
    var n_stacks = (lines[0].length + 1) / 4
    var stacks = (0 ..< n_stacks).map { _ -> ArrayList<Char>() }

    var k = 0
    for (line in lines) {
        var chars = line.asSequence().toList()
        if (chars[1] == '1') {
            break
        }
        k++

        for (i in 0 ..< n_stacks) {
            var name = chars[1 + i * 4]
            if (name != ' ') {
                stacks.get(i).add(name)
            }
        }
    }

    for (line in lines.subList(k + 2, lines.size)) {
        var split = line.split(" ")
        var from = Integer.parseInt(split[3]) - 1
        var to = Integer.parseInt(split[5]) - 1
        for (i in Integer.parseInt(split[1]) - 1 downTo 0) {
            stacks[to].add(0, stacks[from][i])
            stacks[from].removeAt(i)
        }
    }
    println(stacks.stream().map { x -> x.first() }.asSequence().toList())
}
